### ISSUE TITLE
OOM when loading model in fp16 and fine-tuning with LoRA

### ISSUE THREAD

**Issue by menandro** [oobabooga/text-generation-webui#3657]
Is this an implementation issue with PEFT? Why does loading 4-bit work and not fp16? 
In my understanding, the model should be processing in eval mode meaning no gradient. It is true for quantized models like 4-bit because there is literally no gradient implementation there, but 16-bit is not.

Is my understanding correct? Is there a good LoRA finetuning implementation using 16-bit base model?

**Comment by RandomInternetPreson** [comment]
I know this doesn't directly answer your question, but take a look at my enhancement submission here:

https://github.com/oobabooga/text-generation-webui/issues/3637

It links to the QLoRA paper and shows how to get fp16 quality from quantized models.  I do a lot of fine-tuning, and I load the model using transformers using 4-bit, double quant, NF (normal float), these are the parameters that the paper recommends too.

If you load in your model like this, not only can you produce the same quality LoRA as if you had trained on the 16fp model, but you will have way more vram available and can crank up some of the training settings to get an even better result.

After training you can take your FP16 model and merge it with your LoRA if you want to make sure you are getting the highest quality model as a final result.

**Comment by hypersniper05** [comment]
I have a newbie question about training. 

So I have been training Vicuna 13B 1.5 16K GPTQ 4bit model with model loader GPTQ-for-LLaMa and monkey_patch and my training time has been usually around ~18Hrs with the following settings:

Lora Rank 32
Lora Alpha 128
Batch Size 128
Micro Batch Size 2
Cutoff Length 2048
Dataset with ~5k Q&A samples

I switched to Transformers with the same model but 16Bit version, I set my loader using the above settings mentioned by @RandomInternetPreson , In addition I also modified some parameters:  

Lora Rank 128
Lora Alpha 256
Batch Size 2
Cutoff Length 2048
Same Dataset as before

And now my training time is 445Hrs! Ive cranked up the settings before and it was never this high! I am on a 3080ti 12GB .

I also noticed in the console before that training with the GPTQ 4bit model I had 
Training `llama` model using (q, v, o, gate, up, k, down) projections
Trainable params: 125,173,760 (27.6158 %), All params: 453,268,480 (Model: 328,094,720)

and now 

Training 'llama' model using (q, v, o, gate, up, k, down) projections
Trainable params: 500,695,040 (6.9806 %), All params: 7,172,674,560 (Model: 6,671,979,520)

Is that how much is lost because of quantizing? And is this the reason why?

**Comment by hypersniper05** [comment]
Thanks for the reply, yeah I have the options set on the Transformer, I am able to inference the model no problem in for 4bit. 

I did notice like you said the shared memory on my GPU. Since I have 32GB ram it looks like it's offloading some to it. I do have another gpu a 2070 so I might install it. Does the training code automatically split it or do we have to set something up? With Exllama we can specify the gpu's. 

Here are the links :
[https://huggingface.co/TheBloke/vicuna-13B-v1.5-16K-GPTQ](GPTQ)

[https://huggingface.co/lmsys/vicuna-13b-v1.5-16k](Orginal)

**Comment by RandomInternetPreson** [comment]
Interesting, thank you for the information.  It looks like my assumption about GPTQ was incorrect, there are weights removed, however they are removed in such a way as to minimally impact the quality of the resulting model.

https://towardsdatascience.com/4-bit-quantization-with-gptq-36b0f4f02c34

"This method is inspired by a pruning technique to carefully remove weights from a fully trained dense neural network (Optimal Brain Surgeon). It uses an approximation technique and provides explicit formulas for the best single weight w𐞥 to remove and optimal update δꟳ to adjust the set of remaining non-quantized weights F to make up for the removal:"

I'm curious now if the fp16 model using transformers loaded in 4-bit are more accurate than GPTQ, I would suspect so but by what to degree? I would suspect a little bit.

Similarly when you are fine-tuning a GPTQ vs fp16 loaded in with transformers in 4bit, I would suspect there to be a difference but perhaps not much of a difference.

I think your deduction about the trainable percentage of the model is accurate, that more of the model percentage is trained with GPTQ, but there are fewer trainable parameters.  If you look at the total number of trainable parameters between the two methods and the percentage of these parameters that can be changed, they are roughly the same relative percent. 

Without GPTQ there are roughly 4x more trainable parameters and your percentage is roughly 4x less.  So you might be doing very similar manipulations using both methods, with the GPTQ method using less vram because there are fewer parameters to load in.

Regarding your question about multiple GPUs, yup if you can load the model across multiple GPUs (you may or may not need to explicitly dictate this when loading in using transformers, I've found some models need to be told some I can just leave everything at zero MB and the model is loaded evenly among them all) the training will happen across the GPUs without you needing to do anything extra.
